Fix re-migration of existing CA bundles (#21316)
hashicorp/vault: website/content/partials/pki-double-migration-bug.mdx

### PKI Storage Migration Revives Deleted Issuers

> 

Reviewer: ```suggestion
- **Affected versions**: Vault versions before 1.11.1
- **Fixed in**: Vault 1.14.0, 1.13.4, 1.12.8, and 1.11.12
```

Avoid burying the lead ;) Moving the information about which versions are affected, makes it easier for readers to identify the most important piece of information.

> In Vault 1.11, a new storage layout was created to support multiple issuer
> within a single mount, creating storage version 1. In Vault 1.11.6, 1.12.2,
> and 1.13.0, a bug was fixed wherein (due to incorrect ordering of writes),
> the CA chain could be incorrect, due to writes failing when too many mounts
> attempted migration at the same time; this introduced storage version 2.
> 

Reviewer: ```suggestion
Vault 1.11 introduced Storage v1, a new storage layout that supported
multiple issuers within a single mount. Bug fixes in Vault 1.11.6, 1.12.2,
and 1.13.0, corrected a write-ordering issue that lead to invalid CA chains.
Specifically, incorrectly ordered writes would fail and trigger a mass mount
migration. Too many mounts attempting migration at the same time would result
in an incorrect CA chain. The collection of bug fixes introduced Storage v2.
```

Edited for accessibility (active voice, shorter sentences)

> #### Impacted Versions

Reviewer: ```suggestion
```

I would remove the extra heading. It doesn't add clarity and tbh, the text that follows seems more like a continuation of the issue description than a discussion of version-specific impact.

Author: @schavis All of these known issues have `impacted version` headings, so I think we should keep it for consistency as I think users skim it?

Author: ```suggestion
#### Affected versions
```

> If issuers were present pre-1.11, and were subsequently deleted prior to
> upgrading to a Vault version with storage version 2, the initial migration
> of issuers would be re-attempted and these issuers would be recreated. If
> managed keys were associated with these legacy versions (and were
> subsequently removed from the managed key repository), migration would fail
> with an error in the logs:

Reviewer: ```suggestion
Issuers saved under Storage v1 that were deleted before upgrading to a Vault
version with Storage v2 were incorrectly recreated during the upgrade when
Vault automatically migrated legacy issuers. Migration failed if Vault
found managed keys associated with the legacy issuers that had been removed
from the managed key repository. The error appeared in Vault logs as:
```

Edited for accessibility (active voice, shorter sentences)

Author: I think this one is actually incorrect... 

```suggestion
Issuers created before Vault 1.11 and migrated to Storage v1 and deleted
before upgrading to a Vault version with Storage v2 were incorrectly
recreated during the upgrade when Vault automatically re-migrated legacy
issuers. Migration failed if Vault found managed keys associated with the
legacy issuers that had been removed from the managed key repository. The
error appeared in Vault logs as:
```

> Error during migration of PKI mount:
>   failed to lookup public key from managed key:
>     no managed key found with uuid

> This has been fixed in Vault 1.14.0, 1.13.4, 1.12.8, and 1.11.12.

Reviewer: ```suggestion
```

Suggest deleting this line and making the information easier to find by highlighting it at the beginning.

### PKI storage migration revives deleted issuers

Vault 1.11 introduced Storage v1, a new storage layout that supported
multiple issuers within a single mount. Bug fixes in Vault 1.11.6, 1.12.2,
and 1.13.0 corrected a write-ordering issue that lead to invalid CA chains.
Specifically, incorrectly ordered writes could fail due to load, resulting
in the mount being re-migrated next time it was loaded or silently
truncating CA chains. This collection of bug fixes introduced Storage v2.

#### Impacted Versions

> Issuers created before Vault 1.11, migrated to Storage v1, and deleted
> before upgrading to a Vault version with Storage v2 were incorrectly
> recreated during the upgrade when Vault automatically re-migrated legacy
> issuers. Migration could fail if Vault found managed keys associated with the
> legacy issuers that had been removed from the managed key repository. The
> error appeared in Vault logs as:
> 
> > Error during migration of PKI mount:
> >   failed to lookup public key from managed key:
> >     no managed key found with uuid
> 
> Issuers created in Vault 1.11 and later are not impcated. Direct upgrades
> to a Storage v2 layout are not affected.

Reviewer: ```suggestion
Vault may incorrectly re-migrated legacy issuers created before Vault 1.11 that
were migrated to Storage v1 and deleted before upgrading to a Vault version with
Storage v2.

The migration fails when Vault finds managed keys associated with the legacy
issuers that were removed from the managed key repository prior to the upgrade.

The migration error appears in Vault logs as:

> Error during migration of PKI mount:
>   failed to lookup public key from managed key:
>     no managed key found with uuid

<Note>
Issuers created in Vault 1.11+ and direct upgrades to a Storage v2 layout are
not affected.
</Note>
```
style issue: active voice, present tense

Author: Ah, cool, didn't know `<Note>` could be used, I usually use `~> Note: ...` instead. :-)
